### PR TITLE
test: add sovereign to consumer changeover e2e

### PR DESCRIPTION
**Pull request by MSalopek**
## Description

Closes: #780 

Adds e2e testing to establish that sovereign->consumer transition maintains transfer channels and overall functions as expected.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Provided a link to the relevant issue or specification
- [x] Reviewed "Files changed" and left comments if necessary 
- [x] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] Confirmed all author checklist items have been addressed
- [ ] Confirmed that this PR does not change production code

### CODE REVIEW

File: tests/e2e/actions.go [hunk 1]
```diff
@@ -805,6 +942,48 @@ func (tr TestRun) addIbcConnectionGorelayer(
 	tr.waitBlocks(action.chainB, 1, 10*time.Second)
 }
 
+type createIbcClientsAction struct {
+	chainA chainID
+	chainB chainID
+}
+
+func (tr TestRun) createIbcClientsHermes(
```
Comment (MSalopek) [inline tests/e2e/actions.go:950]: When you run `hermes create connection --a-chain <a> --b-chain <b>` hermes will reate NEW clients for you.

File: tests/e2e/actions.go [hunk 2]
```diff
@@ -936,14 +1116,20 @@ func (tr TestRun) addIbcChannelHermes(
...
+
 	//#nosec G204 -- Bypass linter warning for spawning subprocess with cmd arguments.
 	cmd := exec.Command("docker", "exec", tr.containerConfig.instanceName, "hermes",
 		"create", "channel",
 		"--a-chain", string(tr.chainConfigs[action.chainA].chainId),
 		"--a-connection", "connection-"+fmt.Sprint(action.connectionA),
 		"--a-port", action.portA,
 		"--b-port", action.portB,
-		"--channel-version", tr.containerConfig.ccvVersion,
+		"--channel-version", chanVersion,
```
Comment (MSalopek) [inline tests/e2e/actions.go:1132]: making this change to accomodate the fact that the func is used to create IBC `transfer` channels which have a versioning schema (`ICS20-<chan-number>`)

File: tests/e2e/steps_sovereign_changeover.go [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+				chain: chainID("sover"),
+				block: 110,
+			},
+			state: State{},
+		},
+	}
+}
+
+// stepsPostChangeoverDelegate tests basic delegation and resulting validator power changes
+func stepsPostChangeoverDelegate(consumerName string) []Step {
```
Comment (p-offtermatt) [inline tests/e2e/steps_sovereign_changeover.go:235]: Comment why you cannot reuse stepsDelegate

File: tests/e2e/upgrade.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
+package main
```
Comment (p-offtermatt) [inline tests/e2e/upgrade.go:1]: Rename the file to actions_upgrade.go

**Review by p-offtermatt — CHANGES_REQUESTED** [review]
Great work!

Did a joint review with @MSalopek 

Request: * Push sovereign app
* Some minor comments

File: tests/e2e/testnet-scripts/start-changeover.sh [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+# CREATE VALIDATORS AND DO GENESIS CEREMONY
+# !!!!!!!!!!!!!! IMPORTANT !!!!!!!!!!!!!!!! #
+# data dir from the sovereign chain is copied to other nodes (namely bob and carol)
+# alice simply performs an upgrade
+echo "killing nodes"
+pkill -f "^"interchain-security-sd &> /dev/null || true
+
+mkdir -p /root/.sovereign/config
+
+cp -r /sover/validatoralice /.sovereign  # see: app/consumer-democracy/app.go::657
```
Comment (sainoe) [inline tests/e2e/testnet-scripts/start-changeover.sh:43]: Is `/.sovereign` a different directory than `/root/.sovereign`?

File: tests/e2e/actions.go (hunk 1 above)
Comment (MSalopek) [reply tests/e2e/actions.go:950]: Done

File: tests/e2e/actions.go [hunk 6]
```diff
@@ -565,6 +568,139 @@ func (tr TestRun) startConsumerChain(
...
+		}
+		if out == done {
+			break
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		log.Fatal("startChangeover died", err)
+	}
+
+	// tr.addChainToRelayer(addChainToRelayerAction{
```
[the remark this answers is not in the document]
Comment (MSalopek) [reply tests/e2e/actions.go:697]: Done

File: tests/e2e/actions.go [hunk 7]
```diff
@@ -683,6 +819,7 @@ func (tr TestRun) addChainToHermes(
 		keyName,
 		rpcAddr,
 		wsAddr,
+		// action.consumer,
```
[the remark this answers is not in the document]
Comment (MSalopek) [reply tests/e2e/actions.go:822]: Done

File: tests/e2e/config.go [hunk 8]
```diff
@@ -328,6 +331,51 @@ func MultiConsumerTestRun() TestRun {
...
+				binaryName:     "interchain-security-sd",
+				upgradeBinary:  "interchain-security-cdd",
+				ipPrefix:       "7.7.8",
+				votingWaitTime: 20,
+				genesisChanges: ".app_state.gov.voting_params.voting_period = \"20s\" | " +
+					".app_state.slashing.params.signed_blocks_window = \"15\" | " +
+					".app_state.slashing.params.min_signed_per_window = \"0.500000000000000000\" | " +
+					".app_state.slashing.params.downtime_jail_duration = \"2s\" | " +
+					".app_state.slashing.params.slash_fraction_downtime = \"0.010000000000000000\" | " +
+					".app_state.staking.params.unbonding_time = \"1728000s\"",
```
[the remark this answers is not in the document]
Comment (MSalopek) [reply tests/e2e/config.go:371]: Done

File: tests/e2e/steps_sovereign_changeover.go [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+
+// start sovereign chain with 2 validators
+// nodes will cease being validators once the changeover occurs
+func stepRunSovereignChain() []Step {
+	return []Step{
+		{
+			action: StartSovereignChainAction{
+				chain: chainID("sover"),
+				validators: []StartChainValidator{
+					{id: validatorID("alice"), stake: 500000000, allocation: 10000000000},
```
[the remark this answers is not in the document]
Comment (MSalopek) [reply tests/e2e/steps_sovereign_changeover.go:148]: Done

File: tests/e2e/steps_sovereign_changeover.go [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+						// uses val powers from consumer
+						validatorID("alice"): 500,
+						validatorID("bob"):   500,
+						validatorID("carol"): 500,
+					},
+				},
+			},
+		},
+		{
+			action: addIbcConnectionAction{
```
[the remark this answers is not in the document]
Comment (MSalopek) [reply tests/e2e/steps_sovereign_changeover.go:116]: Done

File: tests/e2e/state.go [hunk 11]
```diff
@@ -691,7 +726,14 @@ func (tr TestRun) getQueryNode(chain chainID) string {
 }
 
 // getQueryNodeIP returns query node IP for chain,
-// ipSuffix is hardcoded to be 253 on all query nodes.
+// ipSuffix is hardcoded to be 253 on all query nodes
+// except for "sover" chain where there's only one node
 func (tr TestRun) getQueryNodeIP(chain chainID) string {
+	if chain == chainID("sover") {
+		// return address of first and only validator
```
[the remark this answers is not in the document]
Comment (MSalopek) [reply tests/e2e/state.go:733]: Done

File: tests/e2e/steps_sovereign_changeover.go [hunk 12]
```diff
@@ -0,0 +1,353 @@
+package main
+
+import clienttypes "github.com/cosmos/ibc-go/v4/modules/core/02-client/types"
+
+func stepsSovereignTransferChan() []Step {
+	return []Step{
+		{
+			action: createIbcClientsAction{
```
[the remark this answers is not in the document]
Comment (MSalopek) [reply tests/e2e/steps_sovereign_changeover.go:8]: Done

File: tests/e2e/steps_sovereign_changeover.go (hunk 3 above)
Comment (MSalopek) [reply tests/e2e/steps_sovereign_changeover.go:235]: Done

File: tests/e2e/upgrade.go (hunk 4 above)
Comment (MSalopek) [reply tests/e2e/upgrade.go:1]: Renamed to `actions_sovereign_chain.go`

File: tests/e2e/testnet-scripts/start-changeover.sh (hunk 5 above)
Comment (MSalopek) [reply tests/e2e/testnet-scripts/start-changeover.sh:43]: True, it.
The copy step is not relevant so I removed it.

File: tests/e2e/steps_sovereign_changeover.go [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+func stepsChangeoverToConsumer(consumerName string) []Step {
+	s := []Step{
+		{
+			action: submitConsumerAdditionProposalAction{
+				preCCV:              true,
+				chain:               chainID("provi"),
+				from:                validatorID("alice"),
+				deposit:             10000001,
+				consumerChain:       chainID(consumerName),
+				distributionChannel: "channel-0", // consumer's channel end
```
[the remark this answers is not in the document]
Comment (MSalopek) [reply tests/e2e/steps_sovereign_changeover.go:39]: Done. Comments added

**Review by p-offtermatt — APPROVED** [review]
LGTM, I looked at the comments made after our joint review. I also diffed the app.go to the consumer-democracy one and things looked sane.

**Review by sainoe — APPROVED** [review]
Doesn't have much context here but LGTM. Give my approval since this PR has already been thoroughly reviewed.

**Pull request merged**